branchsearch.cpp

```diff
@@ -93,6 +93,40 @@ void filter_branches(vector<string> &all_branches, char* regex_value, vector<str
     }
 }
 
+bool best_match(vector<string> &all_branches, char* pattern, string& res){
+    // pattern to lowercase
+    std::transform(pattern, pattern+strlen(pattern), pattern, [](unsigned char c){ return std::tolower(c); });
+
+    // take the ones containing the pattern
+    vector<string*> matches;
+    for (string &b: all_branches){
+        // branch to lowercase
+        std::transform(b.begin(), b.end(), b.begin(), [](unsigned char c){ return std::tolower(c); });
+        if (b.find(pattern) != string::npos) matches.push_back(&b);
+    }
+
+    if(!matches.size()) return false;
+
+    // sort by some similarity criteria
+    sort(matches.begin(),matches.end(), [pattern]( const string* a, const string* b)
+        {
+            // the one starting with the pattern is prefered
+            if(a->find(pattern) == 0 and b->find(pattern) != 0) return true;
+            if(b->find(pattern) == 0 and a->find(pattern) != 0) return false;
+
+            // the one closer in length to the pattern is preferd
+            if(a->size()-strlen(pattern) < b->size()-strlen(pattern)) return true;
+            if(b->size()-strlen(pattern) < a->size()-strlen(pattern)) return false;
```

Comment:
Doesn't this consider stuff that doesn't even have the pattern? I'm probably missing something.

Reply:
They were filtered before. Check line 105

Reply:
You're right. My bad.